### ISSUE TITLE
[Azure] Non-interactive way of authenticating with Azure is needed

### ISSUE THREAD

**Issue by ppadala** [docker/machine#3199]
The new Azure driver uses the device login authentication method, which requires the user to enter a code in a browser and then login with Microsoft OneLogin credentials. Though, this may work for interactive users, scripts or management platforms automating management of Azure VMs using docker-machine need a way to automate this.

@ahmetalpbalkan and I discussed this a little bit at https://github.com/docker/machine/pull/3159#discussion_r55563081. Filing this to track it.

P.S. I also noticed that every time docker-machine is run, the user is asked to enter the code. Can this be avoided by entering only once and caching it for certain number of days (I thought that's what docs suggest)?

**Comment by ahmetalpbalkan** [comment]
@ppadala thanks for filing the issue. We are internally looking for a clean way to provide a credentials file all our CLIs (including docker-machine driver) and SDKs can share. There’s currently no ETA for now, however my intention is to find a solution to this in the short term.

> I also noticed that every time docker-machine is run, the user is asked to enter the code. Can this be avoided by entering only once and caching it for certain number of days

This is actually a bug. I'm currently debugging this with one of users. Normally we refresh the token as it expires (every 1 hour) and we can do this for 15 days, and then user is asked to re-authenticate after that time. So it is a manual operation that ought to happen every 2 weeks.

In your case, are you getting asked to authenticate on every single time you run docker-machine? If so, please open a separate issue, I’d love to debug that.

**Comment by ppadala** [comment]
> In your case, are you getting the authentication prompt every single time you run docker-machine? If so, please open a separate issue, I’d love to debug that.

I am running this on Mac, the browser doesn't open automatically, but I get this message. 

Pradeeps-MacBook-Air-2:machine ppadala$ ./bin/docker-machine -D create -d azure --azure-subscription-id $(cat ~/azure/subid) --azure-os "windows" --azure-image a699494373c04fc0bc8f2bb1389d6106__WindowsServerCore_en-us_TP4_Container_Azure-20151118 vhost-win-test
Docker Machine Version:  0.7.0-dev, build a38276a
Found binary path at ./bin/docker-machine
Launching plugin server for driver azure
Plugin server listening at address 127.0.0.1:57286
() Calling .GetVersion
Using API Version  1
() Calling .SetConfigRaw
() Calling .GetMachineName
(flag-lookup) Calling .GetMachineName
(flag-lookup) Calling .DriverName
(flag-lookup) Calling .GetCreateFlags
Found binary path at ./bin/docker-machine
Launching plugin server for driver azure
Plugin server listening at address 127.0.0.1:57290
() Calling .GetVersion
Using API Version  1
() Calling .SetConfigRaw
() Calling .GetMachineName
(vhost-win-test) Calling .GetMachineName
(vhost-win-test) Calling .DriverName
(vhost-win-test) Calling .GetCreateFlags
(vhost-win-test) Calling .SetConfigFromFlags
Running pre-create checks...
(vhost-win-test) Calling .PreCreateCheck
(vhost-win-test) DBG | Set configuration from flags.
(vhost-win-test) DBG | Looking up AAD Tenant ID.  subs="9533c62a-37de-4616-93d0-2ecc89f9ec38"
(vhost-win-test) DBG | Azure request  request="https://management.azure.com/subscriptions/9533c62a-37de-4616-93d0-2ecc89f9ec38?api-version=2015-11-01" method="GET"
(vhost-win-test) DBG | Azure response  request="https://management.azure.com/subscriptions/9533c62a-37de-4616-93d0-2ecc89f9ec38?api-version=2015-11-01" x-ms-request-id="8b19d321-7f04-468d-b939-314d0b0a6b0e" status="401 Unauthorized" method="GET"
(vhost-win-test) DBG | Azure response  request="https://management.azure.com/subscriptions/9533c62a-37de-4616-93d0-2ecc89f9ec38?api-version=2015-11-01" x-ms-request-id="8b19d321-7f04-468d-b939-314d0b0a6b0e" status="401 Unauthorized" method="GET"
(vhost-win-test) DBG | Found AAD Tenant ID.  subs="9533c62a-37de-4616-93d0-2ecc89f9ec38" tenant="ac9085bc-c03c-4ac2-962f-4e0998d24767"
(vhost-win-test) DBG | Loading auth token from file  path="/Users/ppadala/.docker/machine/credentials/azure/ac9085bc-c03c-4ac2-962f-4e0998d24767.json"
(vhost-win-test) DBG | Auth token found in file.  path="/Users/ppadala/.docker/machine/credentials/azure/ac9085bc-c03c-4ac2-962f-4e0998d24767.json"
(vhost-win-test) DBG | Validating the token.
(vhost-win-test) Stored Azure credentials expired. Please reauthenticate.
(vhost-win-test) DBG | Error: Token validity check failed: autorest:WithErrorUnlessStatusCode 400 POST https://login.microsoftonline.com/ac9085bc-c03c-4ac2-962f-4e0998d24767/oauth2/token?api-version=1.0 failed with 400 Bad Request
(vhost-win-test) DBG | Deleting /Users/ppadala/.docker/machine/credentials/azure/ac9085bc-c03c-4ac2-962f-4e0998d24767.json
(vhost-win-test) DBG | Initiating device flow.  path="/Users/ppadala/.docker/machine/credentials/azure/ac9085bc-c03c-4ac2-962f-4e0998d24767.json"
(vhost-win-test) DBG | Azure request  method="POST" request="https://login.microsoftonline.com/ac9085bc-c03c-4ac2-962f-4e0998d24767/oauth2/devicecode?api-version=1.0"
(vhost-win-test) DBG | Azure response  request="https://login.microsoftonline.com/ac9085bc-c03c-4ac2-962f-4e0998d24767/oauth2/devicecode?api-version=1.0" x-ms-request-id="9621b8f8-01c4-4a9b-abe1-047887a46714" status="200 OK" method="POST"
(vhost-win-test) DBG | Retrieved device code.  expires_in=900 interval=5
(vhost-win-test) Microsoft Azure: To sign in, use a web browser to open the page https://aka.ms/devicelogin. Enter the code B65HAN2UM to authenticate.
(vhost-win-test) DBG | Azure request  method="POST" request="https://login.microsoftonline.com/ac9085bc-c03c-4ac2-962f-4e0998d24767/oauth2/token?api-version=1.0"
(vhost-win-test) DBG | Azure response  x-ms-request-id="39793ffc-d3fa-401d-862d-5f414744e269" status="400 Bad Request" method="POST" request="https://login.microsoftonline.com/ac9085bc-c03c-4ac2-962f-4e0998d24767/oauth2/token?api-version=1.0"
(vhost-win-test) DBG | Azure request  method="POST" request="https://login.microsoftonline.com/ac9085bc-c03c-4ac2-962f-4e0998d24767/oauth2/token?api-version=1.0"
(vhost-win-test) DBG | Azure response  method="POST" request="https://login.microsoftonline.com/ac9085bc-c03c-4ac2-962f-4e0998d24767/oauth2/token?api-version=1.0" x-ms-request-id="d8b00211-0aeb-44af-9b66-94643b3bc021" status="400 Bad Request"
(vhost-win-test) DBG | Azure request  method="POST" request="https://login.microsoftonline.com/ac9085bc-c03c-4ac2-962f-4e0998d24767/oauth2/token?api-version=1.0"

After a few login from a browser, this continues. Note that I already applied some of my Windows changes, so not sure if that's causing this. I'll check with unmodified version later.

**Comment by ahmetalpbalkan** [comment]
@ppadala yep, my question was, do you get the `Stored Azure credentials expired. Please reauthenticate.` message every time you run docker-engine? This looks exactly like the issue we're debugging with the other user.

Can you please [use this binary](http://cl.ly/2H0V2g3w133v) (spits out response body for the request that fails just before `credentials expired` message) and send the output line starts like `RESPONSE BODY:` next time it happens? We can't reproduce this with our personal or Active Directory accounts, it is strange.

**Comment by ppadala** [comment]
> @ppadala yep, my question was, do you get the Stored Azure credentials expired. Please reauthenticate. message every time you run docker-engine? This looks exactly like the issue we're debugging with the other user.

Yes, you can see that in the output I pasted. It happens every time I run it, not after 1 hour.

> Can you please use this binary (spits out response body for the request that fails just before credentials expired message) and send the output line starts like RESPONSE BODY: next time it happens? We can't reproduce this with our personal or Active Directory accounts, it is strange.

Ok, will do a bit later.

**Comment by ahmetalpbalkan** [comment]
@ppadala ok let's open a separate issue as we're going off-topic.

**Comment by ahmetalpbalkan** [comment]
@nathanleclaire let's keep this one open, I think the request of authenticating without a browser and having never-expiring credentials is for now a valid use case that we might accomplish in the future. Internal discussions to make this happen are going on.

**Comment by nathanleclaire** [comment]
Sure.

**Comment by vespakoen** [comment]
Yes, this is needed very much, especially for CI / auto deployment.

**Comment by ahmetalpbalkan** [comment]
Hello folks, I implemented a patch at #3439 for Azure driver that accepts Service Principal accounts that eliminate the need for browser-based auth and provide long-lived credentials that are good for headless systems like CI/CD. Feel free to try it out and give feedback.

@ppadala @vespakoen

**Comment by ppadala** [comment]
Thanks @ahmetalpbalkan, much needed feature. To be clear, it's this pull request right? https://github.com/docker/machine/pull/3665

**Comment by ahmetalpbalkan** [comment]
@ppadala ah true.
